### PR TITLE
feat: enable `# fmt: off` recognized by snakefmt

### LINKED ISSUE #86: introduce `# fmt: off` and `# fmt: on` tags to skip formatting within files
This is a follow-up feature request to issue #85 that deals with comment formatting. Handling of [one of the test cases there](https://github.com/snakemake/snakefmt/issues/85#issuecomment-730275169), with aligned comments across multiple lines to hide these lines during docs rendering, [cannot be implemented easily within snakefmt](https://github.com/snakemake/snakefmt/issues/85#issuecomment-736448446).

[As suggested in that issue's discussion](https://github.com/snakemake/snakefmt/issues/85#issuecomment-736512666), one workaround could be comment tags before and after respective sections that turn formatting off and then back on.

[`black` has this](https://github.com/psf/black#the-black-code-style) and uses `# fmt: off` and `#fmt: on`, requiring both comments to be on the same indentation level. I think it would make sense to use exactly the same syntax, so that black users will already know it. And the naming makes perfect sense for `snakefmt`, as well.
In black, you should find the relevant code by searching for `FMT_OFF` in:
https://github.com/psf/black/blob/dea81b7ad5cfa04c3572771c34af823449d0a8f3/src/black/__init__.py

### PR DESCRIPTION
**Pull request by Hocnonsense**
### Format Directives

`snakefmt` supports inline comment directives to control formatting behaviour for specific regions of code.

#### `# fmt: off` / `# fmt: on`

Disables all formatting for the region between the two directives. The directives must appear at the same indentation level. A `# fmt: on` at a deeper indent than the matching `# fmt: off` has no effect.

```python
rule a:
    input:
        "a.txt",


# fmt: off
rule b:
  input: "b.txt"
  output:
          "c.txt"
# fmt: on


rule c:
    input:
        "d.txt",
```

Note: inside `run:` blocks and other Python code, `# fmt: off` / `# fmt: on` is passed through to [Black][black] which handles it natively.

#### `# fmt: off[sort]`

Disables only directive sorting for the region, while still applying all other formatting. Useful when you want to preserve a custom directive order for a specific rule.

```python
# fmt: off[sort]
rule keep_my_order:
    output:
        "result.txt",
    input:
        "source.txt",
    shell:
        "cp {input} {output}"
# fmt: on[sort]
```

A plain `# fmt: on` (without `[sort]`) also ends a `# fmt: off[sort]` region.

#### `# fmt: off[next]`

Disables formatting for the single next Snakemake keyword block (e.g. `rule`, `checkpoint`, `use rule`). Only that one block is left unformatted; subsequent blocks are formatted normally.

```python
rule formatted:
    input:
        "a.txt",
    output:
        "b.txt",


# fmt: off[next]
rule unformatted:
  input: "a.txt"
  output: "b.txt"


rule also_formatted:
    input:
        "a.txt",
```

## Summary by CodeRabbit

* **New Features**
  * Added inline format directives to mark verbatim regions, scoped sorting, and next-line behavior.

* **Documentation**
  * Added a Format Directives section with examples; updated Table of Contents, headings, section names, and Recent Changes formatting.

* **Bug Fixes**
  * Improved directive recognition, comment/indent handling, keyword-boundary detection, and end-of-file/blank-line preservation.

* **Tests**
  * Expanded test suites to cover directives, sorting, boundaries, run-block interactions, and related edge cases.

### CODE REVIEW

File: snakefmt/parser/parser.py [hunk 1]
```diff
@@ -112,13 +171,65 @@ def __init__(self, snakefile: Snakefile):
...
+                    elif "next" in fmt_label.modifiers:
+                        self.fmt_off = (status.cur_indent, "next")
+                        self.fmt_off_expected_index = status.token.line[
+                            : col_nb(status.token)
+                        ]
+                    elif "sort" in fmt_label.modifiers:
+                        self.fmt_sort_off = status.cur_indent
+                elif self._check_fmt_on(status.token) == "sort":
+                    self.fmt_sort_off = None
+                    continue
```
[the remark this answers is not in the document]
Comment (Hocnonsense) [reply snakefmt/parser/parser.py:225]: The `# fmt: off[sort]` / `# fmt: on[sort]` pair is a file-level directive and intentionally takes precedence over the command-line `sort_directives=False` flag. After a `# fmt: on[sort]` directive, sorting being re-enabled is the expected behavior, not a bug. The sentinel value -1 is correctly overwritten by `# fmt: off[sort]` because the in-file directive is meant to override the global setting.

File: snakefmt/parser/parser.py [hunk 2]
```diff
@@ -15,6 +16,53 @@
...
+        TokenInfo(type=62 (FSTRING_MIDDLE), string='\\ncccccccc\\n',
+                  start=(21, 4), end=(23, 0), line="f'''\\ncccccccc\\n'''\\n")
+        TokenInfo(type=63 (FSTRING_END), string="'''",
+                  start=(23, 0), end=(23, 3), line="'''\\n")
+
+    lines should be split to drop overlapping lines and keep unique ones.
+    """
+    return zip(
+        range(token.start[0], token.end[0] + 1), token.line.splitlines(keepends=True)
+    )
```
[the remark this answers is not in the document]
Comment (Hocnonsense) [reply snakefmt/parser/parser.py:64]: This is `token.line.splitlines(keepends=True)`, truncation is impossible!

File: tests/test_formatter.py [hunk 3]
```diff
@@ -47,58 +48,55 @@ def test_single_param_keyword_stays_on_same_line(self):
...
+    example_shell_newline = (
+        "rule a:\n"
+        f'{TAB * 1}shell: "for i in $(seq 1 5);"\n'
+        f'{TAB * 2}"do echo $i;"\n'
+        f'{TAB * 2}"done"',
+        "rule a:\n"
+        f"{TAB * 1}shell:\n"
+        f'{TAB * 2}"for i in $(seq 1 5);" "do echo $i;" "done"\n',
+    )
+
```
Comment (mbhall88) [inline tests/test_formatter.py:60]: I'd prefer to keep all these example code blocks contained inside the test functions so there is no chance they get separated

File: tests/test_formatter.py [hunk 4]
```diff
@@ -1679,17 +1677,24 @@ def test_storage(self):
...
-        formatter = setup_formatter(snakecode)
-        assert formatter.get_formatted() == snakecode
+        assert setup_formatter(expected).get_formatted() == expected
+        snakecode = (
+            "rule fmt_bug_repro:\n"
+            " run:\n"
+            '  if "something nested":\n'
+            "   pass\n"
+            "  # Comment gets indented\n"
+        )
```
Comment (mbhall88) [inline tests/test_formatter.py]: why was this added? it just copies the test code from above, but removes the explicit visual indenting

File: tests/test_formatter.py [hunk 5]
```diff
@@ -2091,3 +2095,613 @@ def test_index_of_first_docstring_match():
...
+            expected = "# fmt: on\n" + formatted
+            assert setup_formatter(code1).get_formatted() == expected
+            code1 = code + "\n\n# fmt: on\n" + code
+            expected = formatted + "\n\n# fmt: on\n" + formatted
+            assert setup_formatter(code1).get_formatted() == expected
+            code1 = "\n# fmt: off\n" + code + "\n# fmt: on\n" + code
+            expected = "# fmt: off\n" + code + "\n# fmt: on\n" + formatted
+            assert setup_formatter(code1).get_formatted() == expected
+
+    def test_fmt_off_not_on(self):
```
Comment (mbhall88) [inline tests/test_formatter.py:2147]: A more descriptive name would be good here. or even a docstring outlining what is being tested as it isn't obvious that you're testing that `# fmt: on` at a deeper indentation level has no effect.

File: tests/test_formatter.py [hunk 6]
```diff
@@ -2091,3 +2095,613 @@ def test_index_of_first_docstring_match():
...
+        code, formatted = TestSimpleParamFormatting.example_shell_newline
+        formatter = setup_formatter(
+            f"rule:\n"
+            f"    run:\n"
+            f"        # fmt: off\n"
+            f"        x = [ 1,2,3]\n"
+            f"        # fmt: on\n\n"
+            f"sth=1\n"
+            f"{code}"
+        )
```
Comment (mbhall88) [inline tests/test_formatter.py:2218]: please use explicit visual tab here. e.g., `f"{TAB * 2}..."`

File: tests/test_formatter.py [hunk 7]
```diff
@@ -2091,3 +2095,613 @@ def test_index_of_first_docstring_match():
...
+            assert setup_formatter(code1, sort_params=True).get_formatted() == expected
+            code1 = "# fmt: off[sort]\n" + code
+            expected = "# fmt: off[sort]\n" + setup_formatter(code).get_formatted()
+            assert setup_formatter(code1, sort_params=True).get_formatted() == expected
+            code2 = code1 + "\n\n# fmt: on[sort]\n" + code
+            expected2 = expected + "\n\n# fmt: on[sort]\n" + formatted
+            assert setup_formatter(code2, sort_params=True).get_formatted() == expected2
+            code2 = code1 + "\n\n# fmt: on\n" + code
+            expected2 = expected + "\n\n# fmt: on\n" + formatted
+            assert setup_formatter(code2, sort_params=True).get_formatted() == expected2
```
Comment (mbhall88) [inline tests/test_formatter.py:2415]: I find this structure very hard to follow (you've also used this above). I prefer to have the concrete examples right within the test that is testing them to be as explicit as possible so a reviewer can clearly see what is being tested and to confirm the formatting/output is as we expect

File: tests/test_formatter.py [hunk 8]
```diff
@@ -2091,3 +2095,613 @@ def test_index_of_first_docstring_match():
...
+            f"{TAB}resources:\n"
+            f"{TAB*2}mem_mb=100,\n"
+            f"{TAB}params:\n"
+            f"{TAB*2}p=1,\n"
+            f"{TAB}message:\n"
+            f'{TAB*2}"finishing"\n'
+            f"{TAB}shell:\n"
+            f'{TAB*2}"echo done"\n'
+        )
+        assert setup_formatter(code, sort_params=True).get_formatted() == expected
```
Comment (mbhall88) [inline tests/test_formatter.py:2610]: this test is an interesting edge case we should discuss.

I would have thought if you turn off sorting around one directive half way through the rule that you would sort the half above it and the half below it, ensuring the directive(s) that is surrounded by `# fmt: off` remain at the same index within the rule? Because what we have done in this example is essentially still sort it, though not quite.... the way it has sorted in the end is strange as it moved `output` to the second directive from the seventh. 

what was your reasoning for sorting it the way you have here?

**Review by mbhall88 — CHANGES_REQUESTED** [review]
Thanks so much for tackling this @Hocnonsense! And great start.

I have added some comments mostly to the tests as these are the only part I really care about. There are some changes I don't like, such as extracting some of the test cases out of the relevant functions into variables outside the tests. I understand this has been done to be able to reuse them in later tests, but I would rather copy and past them into the later tests so that each test can be read completely in isolation and understand what is being tested and whether the formatting makes sense. I can't do this with the updated structure.

We should also support `# fmt: skip` to skip formatting a single line (see [black's docs](https://black.readthedocs.io/en/stable/usage_and_configuration/the_basics.html#ignoring-sections)).

How does the indenting work in the example where the `# fmt: on` is indented closer to 0 than the previous `# fmt: off`? Or does the unindentation implicitly just turn the formatting back on, so the `# fat: on` has no effect? Might be nice to have a test for this.

There is intermittent test code cases where you aren't using explicit visual tab indication. e.g., `f"{TAB * 1}..."` etc. And also cases where you have `f"{TAB}{TAB}.."` should be changed to `f"{TAB * 2}..."`

File: tests/test_formatter.py (hunk 3 above)
Comment (Hocnonsense) [reply tests/test_formatter.py:60]: Thanks for the feedback, but I'd like to argue for keeping the shared variables.

The `(unformatted, expected)` pairing makes the transformation immediately visible in one place: readers can see exactly what changes without jumping between functions.
The later `fmt: off` tests are also deliberately built on top of the earlier formatting examples: they reuse the same code blocks to show that `fmt: off` suppresses the formatting that would otherwise apply, either original or totally formatted, there is no third posibility.
This inheritance is intentional: if the expected formatting changes, the `fmt: off` tests should change in sync, and the shared variable makes that automatic rather than error-prone.

I'd strongly suggest keeping the current structure, but if you'd like `--force` me to inline them I'll do so.

File: tests/test_formatter.py (hunk 4 above)
Comment (Hocnonsense) [reply tests/test_formatter.py]: Sorry for the side effect on this test. I used this example to test whether the "bad" indentation can be formatted, as it should be kept during a `fmt: off` region, but then forgot to remove it. I think you're right that it's unnecessary, and I'll revert my change to this test.

File: tests/test_formatter.py (hunk 5 above)
Comment (Hocnonsense) [reply tests/test_formatter.py:2147]: Sure, I'll add a docstring. This test actually covers two behaviours: (1) `# fmt: on` at a deeper indentation level than `# fmt: off` has no effect, and (2) `# fmt: off` keeps the rest of the code unformatted until a `# fmt: on` at the same indent level is found. I'll make both explicit in the docstring.

File: tests/test_formatter.py (hunk 7 above)
Comment (Hocnonsense) [reply tests/test_formatter.py:2415]: I understand the concern: jumping to another class to read the test cases does add friction for reviewers.

However, the intent of this test is to assert that `# fmt: off[sort]` works correctly regardless of which sortable code block is used.
The loop over multiple examples is expressing that invariant explicitly.
If inlined, that generalisation gets lost and the test reads as verifying one specific case rather than a general behaviour.

You're right that the structure is hard to follow.
The four asserts are actually progressive:
- the first establishes the baseline,
- the second changes the behaviour,
- and the third and fourth extend it.

Rather than splitting into separate tests and losing that progression, I'll add comments to each assert to make the intent explicit. Would that work for you?
That said, I'm happy to inline if you feel strongly about it.

File: tests/test_formatter.py (hunk 8 above)
Comment (Hocnonsense) [reply tests/test_formatter.py:2610]: Thanks for raising this, you're right that the current behaviour is unintuitive.

My reasoning was to treat the `# fmt: off[sort]` block as a unit and insert it just before its natural sorted position, similar to how a preceding comment stays attached to its directive.
However, now I agree that users would expect the directive to stay in place, with the directives above and below sorted independently.

To upate, I added test and am struggling to solve the implementation to match that behaviour.

**Comment by Hocnonsense** [comment]
@mbhall88 Thanks for the detailed review! Here is a summary of what has been addressed:

- Test structure: I've kept the shared class variables for `example_*` and `sorting_*` cases, as they express the invariant that `# fmt: off` suppresses exactly the formatting that would otherwise apply.
    I've added inline comments to the progressive asserts in `test_fmt_off_sort` to make the intent of each step clear.
    Happy to inline everything if you feel strongly about it.
- `# fmt: skip`: Implemented for plain Python lines, matching Black's behaviour.
    Support for Snakemake directive lines (e.g. input:, output:) is not yet implemented, where the parameter parsing path (ParameterSyntax) would require a separate, more involved change.
    A commented-out test and a note in the README document this limitation.
- `# fmt: on` at deeper indent: The behaviour is now documented in the test_fmt_off_not_on docstring:
    - a `# fmt: on` at a deeper indentation level than the matching `# fmt: off` has no effect;
    - formatting remains disabled until a `# fmt: on` at the same level is found,
    - or until the code dedents past the `# fmt: off` level.
- `# fmt: off[sort]` between directives: Updated both the implementation and the test to match your expected behaviour:
    - the region between `# fmt: off[sort]` and `# fmt: on[sort]` stays in place,
    - with directives above and below sorted independently.
- TAB consistency: Unified to `{TAB * N}` style.

**Comment by mbhall88** [comment]
Great work. Two things from me:

1. I tested out the skip functionality but maybe I am not understanding something?

```python
rule:
    input:
        "test.txt"
    output:
        "test.out"
    params:
        test_param="test",
        long_param=lambda wildcards, input: "this is a really long line that should get formatted {}".format(input[0])  # fmt: skip
    shell:
        "echo {input} > {output}"
```

gets formatted to 

```python
rule:
    input:
        "test.txt",
    output:
        "test.out",
    params:
        test_param="test",
        long_param=lambda wildcards, input: "this is a really long line that should get formatted {}".format(
            input[0]
        ),  # fmt: skip
    shell:
        "echo {input} > {output}"
```

I would expect that long line to remain as it is.

2. do the `# fmt: off` lines have to be on their own separate lines, or can they be inline, say 
```
rule:
    input:  # fmt: off
```
and if they can be inline, do they apple to the line they are on, or just subsequent lines?

**Comment by Hocnonsense** [comment]
@mbhall88 Thanks for reviewing!

1. You've hit exactly the known limitation documented in the README:
   `# fmt: skip` currently only applies to plain Python lines, because those are passed through to Black which handles the directive natively.
   Parameter lines inside Snakemake directives (e.g. `rule:`, `conda:`) go through a separate parsing path (`ParameterSyntax`) that does not yet keep original python lines or honour `# fmt: skip`.
   I'd prefer to land the current partial support now and address directive-level `# fmt: skip` in a follow-up PR (but happy to do it in this PR if you'd rather have it complete).

2. `# fmt: off` and `# fmt: on` must appear on their own line (i.e. as a standalone comment, not inline after code). 
   An inline occurrence such as `input:  # fmt: off` is treated as a plain comment and has no effect.
   This matches Black's behaviour.

File: README.md [hunk 9]
```diff
@@ -313,9 +302,103 @@ This ordering ensures that the directives most frequently used in execution bloc
 
 You can disable this feature using the `--no-sort` flag.
 
+### Format Directives
+
+`snakefmt` supports inline comment directives to control formatting behaviour for specific regions of code.
```
Comment (mbhall88) [inline README.md]: in light of our recent discussion, it might be a bit misleading to say "support inline comment" as the comments need to be on their own line

File: README.md [hunk 10]
```diff
@@ -313,9 +302,103 @@ This ordering ensures that the directives most frequently used in execution bloc
...
+rule also_formatted:
+    input:
+        "a.txt",
+```
+
+#### `# fmt: skip`
+
+`# fmt: skip` preserves a single line exactly as written, without any formatting (see [Black's documentation][black-skip] for details).
+
+> **Note:** `# fmt: skip` is not yet supported for lines containing Snakemake directives (e.g. `input:`, `output:`).
```
Comment (mbhall88) [inline README.md]: it is probably clearer to say it is not yet support within snakemake code such as rules as I read this is as I can use it on the line that `input:` is on, but I can use it on other lines within the rule that don't have a directive

**Review by mbhall88 — CHANGES_REQUESTED** [review]

**Comment by mbhall88** [comment]
>     1. You've hit exactly the known limitation documented in the README:
>        `# fmt: skip` currently only applies to plain Python lines, because those are passed through to Black which handles the directive natively.
>        Parameter lines inside Snakemake directives (e.g. `rule:`, `conda:`) go through a separate parsing path (`ParameterSyntax`) that does not yet keep original python lines or honour `# fmt: skip`.
>        I'd prefer to land the current partial support now and address directive-level `# fmt: skip` in a follow-up PR (but happy to do it in this PR if you'd rather have it complete).

Ah okay, I see. It would be nice to have the full support straight away, but I am happy to merge this PR for now, and then we work on adding skip before the next release? I had hoped we would be able to just pass the comment straight through to black for skip as it is a little more manageable to do it inline rather than the more complex block commenting


>     2. `# fmt: off` and `# fmt: on` must appear on their own line (i.e. as a standalone comment, not inline after code).
>        An inline occurrence such as `input:  # fmt: off` is treated as a plain comment and has no effect.
>        This matches Black's behaviour.

Okay, good to know. Might be good to make this explicit in the README

**Comment by Hocnonsense** [comment]
Thanks! Happy to merge now, and I'm happy to follow up with full directive-level `# fmt: skip` support.

Regarding passing it through to Black: the challenge is that parameter lines inside rules go through `ParameterSyntax` rather than Black directly, so there is no clean passthrough path for now. But agreed it is worth solving properly.

File: tests/test_formatter.py (hunk 8 above)
Comment (mbhall88) [reply tests/test_formatter.py:2610]: have you managed to solve the implementation for this?

File: tests/test_formatter.py (hunk 7 above)
Comment (mbhall88) [reply tests/test_formatter.py:2415]: Hmm I understand your reasoning; and it is good reasoning. Let's try thorough documentation within the rules to make sure it is clear what is being asserted?

File: tests/test_formatter.py (hunk 8 above)
Comment (Hocnonsense) [reply tests/test_formatter.py:2610]: Sure, now it acts more naturelly:
https://github.com/snakemake/snakefmt/blob/972860c9a24d9bdbe4e0164c8f63c48ce340cf62/tests/test_formatter.py#L2475-L2525

File: tests/test_formatter.py (hunk 7 above)
Comment (Hocnonsense) [reply tests/test_formatter.py:2415]: Sure, this is updated. (It seems that github skip updating it because of "Large diffs are not rendered by default"), here are the updates:
https://github.com/snakemake/snakefmt/blob/972860c9a24d9bdbe4e0164c8f63c48ce340cf62/tests/test_formatter.py#L2389-L2415

**Review by mbhall88 — APPROVED** [review]
